src/com/perl5/lang/perl/idea/refactoring/PerlRefactor.java

```diff
@@ -0,0 +1,420 @@
+/*
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.perl5.lang.perl.idea.refactoring;
+
+import com.intellij.refactoring.util.AbstractVariableData;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class PerlRefactor {
+  private final String subName;
+  @SuppressWarnings("unused") private String newMethodCall;
+  @SuppressWarnings("unused") private String newMethod;
+
+  private String codeSelection;
+  private String className = "";
+
+  private final Set<String> hashVariables = new HashSet<>();
+  private final Set<String> arrayVariables = new HashSet<>();
+  private final Set<String> scalarVariables = new HashSet<>();
+  private final Set<String> loopScalarVariables = new HashSet<>();
+  private final Set<String> localScalarVariables = new HashSet<>();
+  private final Set<String> localArrayVariables = new HashSet<>();
+  private final Set<String> localHashVariables = new HashSet<>();
+
+
+  private final ArrayList<String> parameters = new ArrayList<>();
+  private final ArrayList<String> innerVariables = new ArrayList<>();
+  private final ArrayList<String> outerVariables = new ArrayList<>();
+  private String methodParameters;
+  private String methodReturn;
+
+  /**
+   * @return The code selection passed in and modified from refactoring
+   */
+  public String getCodeSelection() {
```

Comment:
I'd expected to get some `Selection` from method with such name. Guess `getSelectedCode` or smth would be better name